File: go/tree/convertBST.go
package tree

func convertBST(root *TreeNode) *TreeNode {
	var stack []*TreeNode
	curr := root
	sum := 0
	for curr != nil || len(stack) != 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Right
		}
		curr = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		sum += curr.Val
		curr.Val = sum
		curr = curr.Left
	}
	return root
}

File: java/src/main/java/io/dure/tree/ConvertBST.java
package io.dure.tree;

import java.util.Deque;
import java.util.LinkedList;

public class ConvertBST {
    public TreeNode convertBST(TreeNode root) {
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode curr = root;
        int sum = 0;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.right;
            }
            curr = stack.pop();
            sum += curr.val;
            curr.val = sum;
            curr = curr.left;
        }
        return root;
    }
}
